llvm/CMakeLists.txt

```diff
@@ -918,6 +918,23 @@ if( MINGW AND NOT "${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang" )
   llvm_replace_compiler_option(CMAKE_CXX_FLAGS_RELEASE "-O3" "-O2")
 endif()
 
+if( SYCL_USE_LIBCXX)
```

Comment:
LLVM project already has an option to build with libc++ instead of libstdc++ - `LLVM_ENABLE_LIBCXX`.
My idea was to use this config to build the whole toolchain with libc++.

I hope we don't need all these changes when we use `LLVM_ENABLE_LIBCXX`.

Reply:
Hi, @bader 
Using LLVM_ENABLE_LIBCXX requires the compiler/linker to support "-stdlib=libc++" option which GNU compiler and linker doesn't meet the requirement. This patch just enables building whole sycl toolchain using LLVM libc++ no matter what compiler or linker is used. I am afraid we have to force to use llvm toolchain including compiler/linker/libc++ to build sycl toolchain if switch to use LLVM_ENABLE_LIBCXX. I refer to this: https://github.com/intel/llvm/blob/sycl/llvm/cmake/modules/HandleLLVMStdlib.cmake#L21
Thanks very much.

Reply:
@jinge90, sorry for the delay.
I need provide more context for my [original comment](https://github.com/intel/llvm/pull/2731#discussion_r559972618).
My concern is that we apply custom compile/link options only for part of the libraries and tools instead of applying them to the whole project. This is implemented by adding new cmake functions and additional cmake customization which I'd like to avoid. With current patch I see a couple of things, which doesn't look like a good approach: 

1. Using custom link options for LLVM Support library only.
2. Having SYCL specific cmake code in root cmake file - llvm/CMakeLists.txt. 

If we can't use `LLVM_ENABLE_LIBCXX` to enforce linking with libc++ for the whole project, can we link with libc++ by setting `CMAKE_CXX_FLAGS` and/or `LINK_FLAGS` for the whole project instead?

Reply:
> @jinge90, sorry for the delay.
> I need provide more context for my [original comment](https://github.com/intel/llvm/pull/2731#discussion_r559972618).
> My concern is that we apply custom compile/link options only for part of the libraries and tools instead of applying them to the whole project. This is implemented by adding new cmake functions and additional cmake customization which I'd like to avoid. With current patch I see a couple of things, which doesn't look like a good approach:
> 
> 1. Using custom link options for LLVM Support library only.
> 2. Having SYCL specific cmake code in root cmake file - llvm/CMakeLists.txt.
> 
> If we can't use `LLVM_ENABLE_LIBCXX` to enforce linking with libc++ for the whole project, can we link with libc++ by setting `CMAKE_CXX_FLAGS` and/or `LINK_FLAGS` for the whole project instead?

Hi, @bader 
Just updated the patch to avoid adding any SYCL specific things into root cmake file. We will add "-nostdinc++ -I ${libcxx include}" into CMAKE_CXX_FLAGS when building SYCL toolchain with libcxx, this is done in configure.py.
However, passing "-lc++ -lc++abi -L${libc++ library path}" to whole project in configure.py seems not to work. I dumped the verbose and found those linker flags have been successfully added in all building commands but there are still a lot of "undefined reference" error reported(reference to libc++, libc++abi symbols). The reason is those libc++ linker flags can't be added at the end of the building command, I think the root cause is: https://stackoverflow.com/questions/9417169/why-does-the-library-linker-flag-sometimes-have-to-go-at-the-end-using-gcc
LLVMSupport invokes a lot of std items and the custom link can resolve all referenced libc++ symbols reference, other libraries depend on it seems not to report any undefined libc++/libc++abi when LLVMSupport is fully linked with libc++/libc++abi.
Thanks very much.

Reply:
What do you think about dropping this PR and linking just runtime library with libc++?
I think this customization might bring more problems, but it's not clear if this patch completely removes dependency on libstdc++. It looks like we can link LLVM as shared libraries with libc++. Right?

Reply:
> What do you think about dropping this PR and linking just runtime library with libc++?
> I think this customization might bring more problems, but it's not clear if this patch completely removes dependency on libstdc++. It looks like we can link LLVM as shared libraries with libc++. Right?

Hi, @bader 
Agree. Instead of supporting linking libc++ for whole project, I think building whole sycl project with llvm-toolchain(clang, lld, libc++) is more reasonable and practical, since we can rely on LLVM_ENABLE_LIBCXX.
I will drop this PR. Thanks very much.